### PR TITLE
CALLDATACOPY: Internal call and refactoring

### LINKED ISSUE #134: `CALLDATACOPY` should read caller's memory
Currently `CALLDATACOPY` is implemented to read its own memory when `not is_root` (without `call_id` specified):

https://github.com/appliedzkp/zkevm-specs/blob/af758f49341fa11f2bc8d47a0e71d5b3eed1de13/src/zkevm_specs/evm/execution/memory_copy.py#L22-L33

We can fix this by extend the field `tx_id` of `CopyToMemoryAuxData` to `src_id`, to contain a `tx_id` when `from_tx`, and `caller_id` when `not from_tx`.

And with a slight change to the code above:

```diff
 for i in range(MAX_COPY_BYTES): 
     if not buffer_reader.read_flag(i): 
         byte = FQ.zero() 
     elif aux.from_tx == 1: 
         byte = instruction.tx_context_lookup( 
-            aux.tx_id, TxContextFieldTag.CallData, aux.src_addr + i
+            aux.src_id, TxContextFieldTag.CallData, aux.src_addr + i
         ) 
     else: 
-        byte = instruction.memory_lookup(RW.Read, aux.src_addr + i)
+        byte = instruction.memory_lookup(RW.Read, aux.src_addr + i, caller_id=aux.src_id)
     buffer_reader.constrain_byte(i, byte) 
     if buffer_reader.has_data(i): 
         instruction.constrain_equal(byte, instruction.memory_lookup(RW.Write, aux.dst_addr + i)) 
```

And `CALLDATACOPY` should also set the `src_id` correctly.

### PR DESCRIPTION
**Pull request by roynalnaruto**
Addresses and closes: https://github.com/appliedzkp/zkevm-specs/issues/134

Specs PR:
- [x] https://github.com/appliedzkp/zkevm-specs/pull/178

This PR:
- Refactors `CALLDATACOPY` based on https://github.com/appliedzkp/zkevm-specs/pull/178
- Tests internal calls to `CALLDATACOPY` where `is_root: false`
- Fixes `get_fixed_tables`

### CODE REVIEW

**Comment by roynalnaruto** [comment]
@han0110 https://github.com/appliedzkp/zkevm-circuits/pull/434#issuecomment-1094672164

**Comment by roynalnaruto** [comment]
@han0110 I've updated the bus-mapping tests for `CALLDATACOPY` to test an internal call case:
https://github.com/scroll-tech/zkevm-circuits/blob/fix/calldatacopy-internal/bus-mapping/src/evm/opcodes/calldatacopy.rs#L201

which passes as expected.

Although the same test fails in the `evm_circuit/execution` module:
https://github.com/scroll-tech/zkevm-circuits/blob/fix/calldatacopy-internal/zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs#L500

with this error:
```
`Err([Lookup { name: "LOOKUP: \"Fixed table\"", lookup_index: 33, location: InRegion { region: Region { index: 5, name: "Execution step" }, offset: 176 } }, Lookup { name: "LOOKUP: \"Fixed table\"", lookup_index: 38, location: InRegion { region: Region { index: 5, name: "Execution step" }, offset: 176 } }, Lookup { name: "LOOKUP: \"Bytecode table\"", lookup_index: 162, location: InRegion { region: Region { index: 5, name: "Execution step" }, offset: 288 } }])`
```

Do you know what might be the issue here?

**Comment by han0110** [comment]
1. I forgot to update the [`get_fixed_table`](https://github.com/appliedzkp/zkevm-circuits/blob/main/zkevm-circuits/src/test_util.rs#L22) to have `Range64` and `Range1024` (my apology), so 2 `fixed_table` lookup failed. Adding them should remove first 2 failures.
2. Currently `STOP` can't handle index bytecode out of range, so the `bytecode_table` lookup failed. Adding a `STOP` after https://github.com/scroll-tech/zkevm-circuits/blob/fix/calldatacopy-internal/zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs#L533 should remove the last failure.

**Comment by roynalnaruto** [comment]
Should I update `get_fixed_table` and commit in this PR?

**Comment by han0110** [comment]
Yes, let's fix it in this PR. Thanks!

File: zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs [hunk 1]
```diff
@@ -488,4 +495,61 @@ mod test {
             Word::from(0),
         );
     }
+
+    #[test]
+    fn calldatacopy_gadget_busmapping_internal() {
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs:500]: you can remove the old testing that manually creates the steps and rw operations.

**Review by icemelon — APPROVED** [review]
Thanks for implementing this. Overall LGTM. Just one minor comment

File: zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs (hunk 1 above)
Comment (roynalnaruto) [reply zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs:500]: Sure. I also need to do multiple tests with different arguments. Will push in a while.

**Comment by roynalnaruto** [comment]
@han0110 This is ready to be reviewed again. All changes included from my end.

**Review by han0110 — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**